<div style="text-align:left;font-size:2em"><span style="font-weight:bolder;font-size:1.25em">SP2273 | Learning Portfolio</span><br><br><span style="font-weight:bold;color:darkred">Loops (Good) Exercises</span></div>

### Exercise 1 (Make me an odd list)

In [20]:
%%timeit 
numbers = []
for i in range(10):
    if i % 2 == 0:
        continue
    numbers.append(i ** 2)
#print(numbers)

1.13 μs ± 37.3 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


### Exercise 2 (Make me another odd list)

In [17]:
%%timeit
[number ** 2 for number in range(10) if number % 2 == 1]

994 ns ± 30.5 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


### Exercise 3 (Time me!)

From exercise 1, the mean time taken per loop is 1.13 μs, with a standard deviation of 37.3ns out of the 7 x 1,000,000 loops done.  
From exercise 2, the mean time taken per loop is 994 ns, with a standard deviation of 30.5ns out of the 7 x 1,000,000 loops done.  

This implies that:  
- Using list comprehension results in faster processing speeds for the same work done from growing lists (994 ns < 1.13 μs)
- Furthermore, list comprehension yields more consistent speeds for the same work done as compared to growing lists (30.5 ns < 37.5 ns)

### Exercise 4 (A problem of decay)

In [33]:
decay = int(input('Initial Quantity of Sample is ___ units?'))
sample = decay * 0.5
day = 0
while True:
    day += 1
    decay *= 0.95
    if decay <= sample: break
print(f'Half-life = {day} days') 

Initial Quantity of Sample is ___ units? 100


Half-life = 14 days


### Exercise 5 (Changes in CO<sub>2</sub>)

In [34]:
co2_data = [
    (2000, 369.55), (2001, 371.14), (2002, 373.28), 
    (2003, 375.80), (2004, 377.52), (2005, 379.80), 
    (2006, 381.90), (2007, 383.79), (2008, 385.60), 
    (2009, 387.43), (2010, 389.90), (2011, 391.65), 
    (2012, 393.85), (2013, 396.52), (2014, 398.65),
    (2015, 400.83), (2016, 404.24), (2017, 406.55), 
    (2018, 408.52), (2019, 411.44), (2020, 414.24)
]

In [69]:
year1 = []
year2 = []
increase = []
a = len(co2_data) 
for i in range(a):
    if co2_data[i][1] - co2_data[i-1][1] < 3:
        continue
    year1.append(co2_data[i-1][0])
    year2.append(co2_data[i][0])
    increase.append(co2_data[i][1] - co2_data[i-1][1])
n = len(year1)
for j in range(n):
    print(f'There is an increase in the level of CO2 in the atmosphere from {year1[j]} to {year2[j]} of {increase[j]:.2f} ppm')


There is an increase in the level of CO2 in the atmosphere from 2015 to 2016 of 3.41 ppm
